File: Medium/Count Occurences of Anagrams/count-occurences-of-anagrams.py
#User function Template for python3
class Solution:
	def search(self,pat, txt):
	    d={}
	    for i in range(len(pat)):
	        d[pat[i]]=d.get(pat[i],0)+1
	    m={}
	    cnt=0
	    k=len(pat)
	    for i in range(k):
	        m[txt[i]]=m.get(txt[i],0)+1
	    if m==d:
	        cnt=cnt+1
	    for i in range(k,len(txt)):
	        m[txt[i-k]]=m.get(txt[i-k],0)-1
	        if m[txt[i-k]]==0:
	            del m[txt[i-k]]
	        m[txt[i]]=m.get(txt[i],0)+1
	        if m==d:
	            cnt=cnt+1
	    return cnt
	        
	    
	   
	   


#{ 
 # Driver Code Starts
#Initial Template for Python 3


# Driver code 
if __name__ == "__main__": 		
    tc=int(input())
    while tc > 0:
        txt=input().strip()
        pat=input().strip()
        ob = Solution()
        ans = ob.search(pat, txt)
        print(ans)
        tc=tc-1
# } Driver Code Ends

File: Difficulty: Medium/Count Occurences of Anagrams/count-occurences-of-anagrams.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    // Driver code
    public static void main(String[] args) throws Exception {
        BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine().trim());
        while (t-- > 0) {
            String txt = br.readLine().trim();
            String pat = br.readLine().trim();

            int ans = new Solution().search(pat, txt);

            System.out.println(ans);
        }
    }
}
// } Driver Code Ends


// User function Template for Java

class Solution {

    int search(String pat, String txt) {
        int []p=new int[26];
        int []t=new int[26];
        int count=0;
        for(int i=0;i<pat.length();i++){
            p[pat.charAt(i)-'a']=p[pat.charAt(i)-'a']+1;
        }
        int k=pat.length();
        for(int i=0;i<k;i++){
            t[txt.charAt(i)-'a']=t[txt.charAt(i)-'a']+1;
        }
        boolean flag=true;
        for(int i=0;i<26;i++){
            if(p[i]!=t[i]){
                flag=false;
                break;
            }
        }
        if(flag==true) count=count+1;
        for(int i=k;i<txt.length();i++){
            t[txt.charAt(i-k)-'a']=t[txt.charAt(i-k)-'a']-1;
            t[txt.charAt(i)-'a']=t[txt.charAt(i)-'a']+1;
            boolean Tflag=true;
            for(int j=0;j<26;j++){
                if(p[j]!=t[j]){
                    Tflag=false;
                    break;
                }
            }
            if(Tflag==true) count=count+1;
        }
        return count;
    }
}